pkg/cfn/manager/tasks.go

```diff
@@ -1,158 +1,208 @@
-		if onlySubset != nil && !onlySubset.Has(ng.Name) {
-			continue
-		}
-		t := Task{
-			Call: c.CreateNodeGroup,
-			Data: ng,
-		}
-		createAllNodeGroups = append(createAllNodeGroups, t)
+	if t.Parallel {
+		doParallelTasks(appendErr, t.tasks)
+	} else {
+		doSequentialTasks(appendErr, t.tasks)
 	}
-	if Run(appendErr, createAllNodeGroups...); len(errs) > 0 {
+	if len(errs) > 0 {
 		return errs
 	}
-
 	return nil
 }
 
-// CreateOneNodeGroup runs a task to create a single node groups;
-// any errors will be returned as a slice as soon as the tasks is
-// completed
-func (c *StackCollection) CreateOneNodeGroup(ng *api.NodeGroup) []error {
-	return c.RunSingleTask(Task{
-		Call: c.CreateNodeGroup,
-		Data: ng,
-	})
+type taskWithoutParams struct {
+	info string
+	call func(chan error) error
 }
 
-// DeleteAllNodeGroups deletes all nodegroups without waiting
-func (c *StackCollection) DeleteAllNodeGroups(call taskFunc) []error {
-	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
-	if err != nil {
-		return []error{err}
-	}
+func (t *taskWithoutParams) Describe() string { return t.info }
+func (t *taskWithoutParams) Do(errs chan error) error {
+	return t.call(errs)
+}
 
-	errs := []error{}
-	for _, s := range nodeGroupStacks {
-		if err := c.DeleteNodeGroup(getNodeGroupName(s)); err != nil {
-			errs = append(errs, err)
-		}
-	}
+type taskWithNameParam struct {
+	info string
+	name string
+	call func(chan error, string) error
+}
 
-	return errs
+func (t *taskWithNameParam) Describe() string { return t.info }
+func (t *taskWithNameParam) Do(errs chan error) error {
+	return t.call(errs, t.name)
 }
 
-// WaitDeleteAllNodeGroups runs all tasks required to delete all the nodegroup
-// stacks and wait for all nodegroups to be deleted; any errors will be returned
-// as a slice as soon as the group of tasks is completed
-func (c *StackCollection) WaitDeleteAllNodeGroups(force bool) []error {
-	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
-	if err != nil {
-		return []error{err}
-	}
+type taskWithNodeGroupSpec struct {
+	info      string
+	nodeGroup *api.NodeGroup
+	call      func(chan error, *api.NodeGroup) error
+}
 
-	errs := []error{}
-	appendErr := func(err error) {
-		errs = append(errs, err)
-	}
+func (t *taskWithNodeGroupSpec) Describe() string { return t.info }
+func (t *taskWithNodeGroupSpec) Do(errs chan error) error {
+	return t.call(errs, t.nodeGroup)
+}
 
-	deleteAllNodeGroups := []Task{}
-	for i := range nodeGroupStacks {
-		t := Task{
-			Call: c.WaitDeleteNodeGroup,
-			Data: getNodeGroupName(nodeGroupStacks[i]),
-		}
-		if force {
-			t.Call = c.WaitForceDeleteNodeGroup
-		}
-		deleteAllNodeGroups = append(deleteAllNodeGroups, t)
+type taskWithStackSpec struct {
+	info  string
+	stack *Stack
+	call  func(*Stack, chan error) error
+}
+
+func (t *taskWithStackSpec) Describe() string { return t.info }
+func (t *taskWithStackSpec) Do(errs chan error) error {
+	return t.call(t.stack, errs)
+}
+
+type asyncTaskWithStackSpec struct {
```

Comment:
can you elaborate what the purpose for this struct is and what is intended to happen when the error channel is closed in `Do()`?

Reply:
So the error channel is closed here right after the handler returns, to indicate the task has completed. Maybe we could give this one a better name... It's a fire-and-forget type of task.